### ISSUE TITLE
Environment stops responding with high CPU usage

### ISSUE THREAD

**Issue by rwightman** [Unity-Technologies/obstacle-tower-env#77]
I've started running into an issue with the environment / game engine. After running the train loop for some time, part of a day or more, the environment appears to stop responding. I haven't dug in too much yet but it appears as though it gets into a state where it doesn't respond to the python side of the step request. 

Once the issue occurs, the training loop is stuck, my Python RL processes are idle with no CPU consumption and the environment/game engine process is using about 5-6x more CPU than during normal function -- consuming 50-60% of all cores on a 6 core 8700K according to htop. I'm running on ubuntu 16.04.

This didn't happen when I had the system training for over a week. I'm wondering if hitting higher levels causes it as I've been pushing into/above floor 10 since I've been having the problem with no noteworthy changes to the mechanics on the Python RL algo side. 

Has anyone else seen this issue or an ideas what it might be or how to debug it without adding a significant slowdown to the system?

**Comment by awjuliani** [comment]
Hi @rwightman 

Thanks for bringing this to our attention. Around the floor 10 point is also where the second visual theme is introduced. There could possibly be an issue related to that. Would it be possible to try saving out some visual observations to see where things hang in your setup?

**Comment by rwightman** [comment]
Yeah, wondering if I could rig up an async thread/timer to fire on the step call not returning and dump part of my replay buffer without upsetting the GIL. Before that though, I'll see if I can attach a debugger cleanly once it's stuck, should happen in the next few hours...

**Comment by rwightman** [comment]
No dice on being able to attach the debugger and see much of anything. But in this instance the issue occured on a different machine. I was trying some new ideas training from scratch, this one was training fine for several days and just died tonight as it started to hit floor 10 on occasion. 

On my other machine, resuming an older model that was close to reaching floor 10 usually hits the issue within a few hours.

**Comment by Holt59** [comment]
Me (and some colleagues) ran into very similar issues. We are running RL algorithms with MPI communications for the agents, and when the agent starts to reach level 10, the environments seem to crash. 

I am running ~60 environments and the problem seems to appear after some hours of training, but one of my colleague ran into the same issue with a single environment.

I found some strange logs (in `Player.log`), which might be related, but it's hard for me to do more debugging since this is kind of random... 


> Internal: JobTempAlloc has allocations that are more than 4 frames old - this is not allowed and likely a leak
> (Filename:  Line: 372)
> To Debug, enable the define: TLA_DEBUG_STACK_LEAK in ThreadsafeLinearAllocator.cpp. This will output the callstacks of the leaked allocations
> (Filename:  Line: 374)

**Comment by Holt59** [comment]
I just launch a run after posting the previous comments (before going to sleep) and it ran... 9 minutes. This is kind of annoying to advance in the competition... 

`Player.log` contains different, here is the whole file: https://gist.github.com/Holt59/1679694fc3dcedbb2e9e8114c06a39bc

**Comment by VilmantasBal** [comment]
Hey,

are you running several Unity processes on the same machine ? While looking at the log 
"rrequesting resize 84 x 84rresizing window to 84 x 84
Desktop is 1280 x 1024 @ 0 HzAAcademy resetting"

it seems that at least two processes are trying concurrently to write to log file ?

Kind regards,
Vilmantas

**Comment by VilmantasBal** [comment]
Anyway, i am trying to repro this on our end. Stay tuned.

**Comment by Holt59** [comment]
@VilmantasBal Yes, as I mentioned I am running multiple environments (~60). 

A colleague of mine had similar problems with a single environment but these were less frequent.

**Comment by rwightman** [comment]
![seq](https://user-images.githubusercontent.com/5702664/55381054-f021eb00-54d6-11e9-88a6-ee26fe4ab0ff.gif)

I wrote an env wrapper that dumps last observations when the step call doesn't return. I attached one sequence that I grabbed, it doesn't look particularly noteworthy. Hmm

**Comment by rwightman** [comment]
If anyone is interested in helping debug (or work around the issue), here is a snippet of the wrapper I wrote.

```
class TimeoutMonitor(gym.Wrapper):
    def __init__(self, env, memory):
        gym.Wrapper.__init__(self, env)
        self._memory = memory
        self._running = True
        self._waiting = False
        self._timeout = 30
        self._cv = threading.Condition()
        self._thread = threading.Thread(target=self._run)
        self._thread.setDaemon(True)
        self._thread.start()

    def step(self, action):
        self._set_waiting()
        result = self.env.step(action)
        self._clear_waiting()
        return result

    def reset(self, **kwargs):
        return self.env.reset(**kwargs)

    def close(self):
        with self._cv:
            self._running = False
            self._waiting = False
            self._cv.notify_all()
        self._thread.join()
        return self.env.close()

    def _set_waiting(self):
        with self._cv:
            assert not self._waiting
            self._waiting = True
            self._cv.notify()

    def _clear_waiting(self):
        with self._cv:
            assert self._waiting
            self._waiting = False
            self._cv.notify()

    def _run(self):
        with self._cv:
            print('Timeout monitor active...')
            while self._running:
                self._cv.wait_for(lambda: self._waiting or not self._running)
                if not self._running:
                    break
                not_expired = self._cv.wait_for(lambda: not self._waiting or not self._running, self._timeout)
                if not not_expired:
                    print('TIMEOUT!')
                    self._dump_memory()
                    raise TimeoutError()

    def _dump_memory(self, n=60):
        pass
```

**Comment by VilmantasBal** [comment]
Hey Ross,

thanks, i ll take a look, this is useful.

Kind regards,
Vilmantas

**Comment by Holt59** [comment]
I already tried with the following:

```python
from concurrent.futures import ThreadPoolExecutor
import concurrent.futures as cfutures

# Warning: this does not terminate function if timeout
def timeout_five(fnc, *args, **kwargs):
    with ThreadPoolExecutor() as p:
        f = p.submit(fnc, *args, **kwargs)
        return f.result(timeout=5)

# In reset / step
try:
    observation = timeout_five(self.environment.reset)
except cfutures._base.TimeoutError:
    print(str(datetime.datetime.now()) + ": WARNING timeout environment.reset")
    with open('LOG_WARNING.log', 'w') as f:
        f.write(str(datetime.datetime.now()) + ": WARNING timeout environment.reset")
```

But I did not get any log, even after the program got stuck. I'll check your version if I can but it can get kind of messy with MPI.

**Comment by VilmantasBal** [comment]
Hey,

can you check memory consumed by Unity process ? And total free memory of
machine ? We are suspecting there is a memory leak of some sort.

Kind regards,
Vilmantas

**Comment by Holt59** [comment]
@VilmantasBal I've launch a new learning process and I'll keep you updated with the results.

However, as I mentioned, this happens kind of "randomly" &mdash; Typically after ~1 day of learning when starting from scratch, and ~2-3 hours of learning when starting from an agent that reaches level 5. But it also sometimes crashes after less than 15 minutes... 

I have ~50GB memory on this VM, and each environment seems to be taking less than 300MB (currently), so unless there is a huge spike at some point, I doubt I'll get memory issues.

**Comment by Holt59** [comment]
@VilmantasBal My learning after ~4 hours, highest memory usage for the environments was ~330MB, so less than 20GB for the 60 environments, far from my 50GB.

I checked the GPUs with `nvidia-smi`, with similar results.

At the end, one of the `obstacletower.x86_64` process was taking about 400% CPUs, while the others were around 0.5-2%. I try to kill the top environment, but nothing changed.

**Comment by rwightman** [comment]
@VilmantasBal If there is a memory leak, it's slight and far from putting any pressure on my system. Same as @Holt59, my environment(s) are around 330-340MB. 

A small memory leak wouldn't be the best explanation for the sudden lock-up and massive increase in CPU consumption when the issue occurs. You'd typically see a gradual decline in performance and responsiveness.

**Comment by VilmantasBal** [comment]
OK, but where is your python training process running ? On the same machine
where Unity env are running ? How much memory consumed is by python
training ? Say in default dopamine rainbow agent implementation with
parameters provided python process can consume more than 7 gb of memory
because of experience replay buffer which stores 1 million of previous
states(frame, action, reward, next frame).

**Comment by Holt59** [comment]
@VilmantasBal I don't know if you're addressing me or @rwightman so I'll answer anyway.

Everything is running on a single host, but I am not using Rainbow, I am using PPO, so no replay buffer, The total memory used on the host is around 18-20 GB when the 60 environments and the learning process are running, and seems to remain stable throughout the learning process. The GPU memory consumption is also quite stable, around 5 GB from the 11 available.

**Comment by rwightman** [comment]
@VilmantasBal Different from @Holt59 in that I'm using a DQN style training loop with replay, and thus more memory in one process but less Python process and only 1-2 environments. The overall situation is the same though, memory usage appears stable enough after everything settles in (which takes quite a while). My system has 64GB of memory and the usage is well within that, about half the system memory is completely free and unused by any user process or OS, so we're far from any memory pressure scenario even if there were a leak. Absolutely nothing is being paged out.

Also, I have spent a lot of time debugging systems over the years. This doesn't behave like an issue caused by a memory leak.

**Comment by VilmantasBal** [comment]
Hey,

one more question about your setup: do you use docker ? Or some kind of VM
? Do you use xfvb instead of normal X server ? Software rasterizer like
llvmpipe instead of hardware accelerated drivers like nVidia proprietary ?


Kind regards,
Vilmantas

**Comment by rwightman** [comment]
@VilmantasBal I'm running on and experienced the issue on two different systems, Ubuntu 16.04 and 18.04, nvidia proprietary drivers, both in an fairly standard X11 Ubuntu MATE desktop setup. Realtime mode and thus rendering to the desktop window is disabled. The rendering for the observations looks to be using the GPU as 89MB of GPU memory is allocated by the environment process.

**Comment by Holt59** [comment]
@VilmantasBal I am running this on a GCloud VM, I followed the tutorial for Dopamine for setting up the VM https://github.com/Unity-Technologies/obstacle-tower-env/blob/master/examples/gcp_training.md.

**Comment by rwightman** [comment]
I have tried attaching GDB to the environment process when the issue happens. There isn't much to go on without symbols but it's always the same. There are 11 threads using significant CPU, one thread using a little, and then the rest completely idle. 

From the instance below, threads 4-14 are all consuming approx 50% of a virtual core' worth of CPU. And 32 is using 5%. If you can provide an debug build of obstacletower.x86_64 or lightly optimized build with symbols I could provide stack traces within a day.

```
  Id   Target Id         Frame 
* 1    Thread 0x7f4fa6c79740 (LWP 2540) "obstacletower.x" 0x00007f4fa6694c0c in futex_wake (private=<optimized out>, processes_to_wake=1, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:231
  2    Thread 0x7f4fa3307700 (LWP 2545) "SDLTimer" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x3a29750) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  3    Thread 0x7f4f9d8b4700 (LWP 2546) "Finalizer" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4f9e1da5b0) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  4    Thread 0x7f4f9451e700 (LWP 2547) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  5    Thread 0x7f4f93d1d700 (LWP 2548) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  6    Thread 0x7f4f9351c700 (LWP 2549) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  7    Thread 0x7f4f92d1b700 (LWP 2550) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  8    Thread 0x7f4f9251a700 (LWP 2551) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  9    Thread 0x7f4f91d19700 (LWP 2552) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  10   Thread 0x7f4f91518700 (LWP 2553) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  11   Thread 0x7f4f90d17700 (LWP 2554) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  12   Thread 0x7f4f73fff700 (LWP 2555) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  13   Thread 0x7f4f737fe700 (LWP 2556) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  14   Thread 0x7f4f72ffd700 (LWP 2557) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  15   Thread 0x7f4f727fc700 (LWP 2558) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  16   Thread 0x7f4f71ffb700 (LWP 2559) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  17   Thread 0x7f4f717fa700 (LWP 2560) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  18   Thread 0x7f4f70ff9700 (LWP 2561) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  19   Thread 0x7f4f4ffff700 (LWP 2562) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  20   Thread 0x7f4f4f7fe700 (LWP 2563) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  21   Thread 0x7f4f4effd700 (LWP 2564) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  22   Thread 0x7f4f4e7fc700 (LWP 2565) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  23   Thread 0x7f4f4dffb700 (LWP 2566) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  24   Thread 0x7f4f4d7fa700 (LWP 2567) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  25   Thread 0x7f4f4cff9700 (LWP 2568) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  26   Thread 0x7f4f2ffff700 (LWP 2569) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  27   Thread 0x7f4f2f7fe700 (LWP 2570) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  28   Thread 0x7f4f2effd700 (LWP 2571) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  29   Thread 0x7f4f2e7fc700 (LWP 2572) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  30   Thread 0x7f4f2dffb700 (LWP 2573) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d128a8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  31   Thread 0x7f4f2d7fa700 (LWP 2574) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa490ecd8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  32   Thread 0x7f4f17fff700 (LWP 2575) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa548dd28) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  33   Thread 0x7f4f16c98700 (LWP 2577) "threaded-ml" 0x00007f4fa590d74d in poll () at ../sysdeps/unix/syscall-template.S:84
  34   Thread 0x7f4f9e72d700 (LWP 2578) "obstacletower.x" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:143
  35   Thread 0x7f4f9c0cf700 (LWP 2579) "obstacletower.x" 0x00007f4fa58de30d in nanosleep () at ../sysdeps/unix/syscall-template.S:84
  36   Thread 0x7f4f9c07e700 (LWP 2580) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d355f8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  37   Thread 0x7f4f06ffe700 (LWP 2583) "obstacletower.x" 0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa4009b78) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  38   Thread 0x7f4f063a3700 (LWP 2584) "grpc_executor" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  39   Thread 0x7f4f05ba2700 (LWP 2585) "grpc_global_tim" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  40   Thread 0x7f4f053a1700 (LWP 2586) "grpc 0 (cq 0)" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  41   Thread 0x7f4f051a0700 (LWP 2587) "grpc 1 (cq 1)" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  42   Thread 0x7f4f04f9f700 (LWP 2588) "grpc 2 (cq 2)" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  43   Thread 0x7f4f04d9e700 (LWP 2589) "grpc 3 (cq 3)" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  44   Thread 0x7f4f04b9d700 (LWP 2590) "grpc 4 (cq 4)" 0x00007f4fa5919a13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  45   Thread 0x7f4f0499c700 (LWP 2591) "grpc 5 (cq 5)" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  46   Thread 0x7f4efbdfd700 (LWP 2597) "grpc_global_tim" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  47   Thread 0x7f4f04388700 (LWP 4835) "Thread Pool Wor" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  48   Thread 0x7f4efbffe700 (LWP 4836) "Thread Pool Wor" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  49   Thread 0x7f4fa3cfd700 (LWP 4971) "grpc 0 (cq 0)" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
```

Thread 4 backtrace
```
#0  0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa3d12338) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  do_futex_wait (sem=sem@entry=0x7f4fa3d12338, abstime=0x0) at sem_waitcommon.c:111
#2  0x00007f4fa66948d4 in __new_sem_wait_slow (sem=0x7f4fa3d12338, abstime=0x0) at sem_waitcommon.c:181
#3  0x00007f4fa669497a in __new_sem_wait (sem=<optimized out>) at sem_wait.c:29
#4  0x00000000009cbfe6 in ?? ()
#5  0x00000000009cc08b in ?? ()
#6  0x0000000000a83721 in ?? ()
#7  0x00007f4fa668c6ba in start_thread (arg=0x7f4f9451e700) at pthread_create.c:333
#8  0x00007f4fa591941d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

Thread 32 backtrace
```
#0  0x00007f4fa6694827 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f4fa548dd28) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  do_futex_wait (sem=sem@entry=0x7f4fa548dd28, abstime=0x0) at sem_waitcommon.c:111
#2  0x00007f4fa66948d4 in __new_sem_wait_slow (sem=0x7f4fa548dd28, abstime=0x0) at sem_waitcommon.c:181
#3  0x00007f4fa669497a in __new_sem_wait (sem=<optimized out>) at sem_wait.c:29
#4  0x0000000000a852b9 in ?? ()
#5  0x0000000000bbd0e8 in ?? ()
#6  0x0000000000bbb6c2 in ?? ()
#7  0x0000000000bbb7e9 in ?? ()
#8  0x0000000000a83721 in ?? ()
#9  0x00007f4fa668c6ba in start_thread (arg=0x7f4f17fff700) at pthread_create.c:333
#10 0x00007f4fa591941d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

**Comment by rwightman** [comment]
@VilmantasBal noteworthy, all the threads chewing through excessive CPU once the env stops responding are waiting on the same futex. Should be easy enough to identify at least which thread those are on your end....

**Comment by rwightman** [comment]
I can also make a core dump available privately if you have symbols to pair with the released exe.

**Comment by unixpickle** [comment]
I managed to reproduce this issue deterministically with a simple script. The script runs a certain set of actions on a certain floor and seed, and it hangs on the last step.

This should make it much easier to debug the issue. It proves that the issue has nothing to do with your training setup--**it's purely a bug in the environment.**

This reproduction works on Ubuntu 18.04, both with the environment running on CPU and GPU. I have not tested on macOS or Windows, since I don't have access to such systems.

Here's the script:

```python
from obstacle_tower_env import ObstacleTowerEnv

ENV_PATH = './ObstacleTower/obstacletower'

ACTIONS = ([18]*6 + [24]*3 + [18]*9 + [24]*2 + [18]*9 + [30]*4 + [24] + [18]*5 + [30, 18, 24] +
           [30]*3 + [24]*2 + [18] + [30]*2 + [24]*2 + [18]*4 + [30]*6 + [24]*4 + [30]*7 + [24]*13 +
           [18] + [30]*3 + [18, 0] + [6]*4 + [24] + [30]*3 + [18]*2 + [0, 21, 0] + [12]*2 + [6]*3 +
           [0]*2 + [21]*2 + [12] + [6]*3 + [12]*3 + [21]*2 + [18, 30] + [6]*3 + [30]*2 + [24] +
           [12]*13 + [30] + [18]*3 + [24]*3 + [21]*2 + [24]*18 + [21] + [30]*2 + [24]*4 + [18]*3 +
           [24] + [18]*2 + [24, 30] + [24, 30] + [24]*10 + [15] + [24]*7 + [18] + [30]*3 + [21] +
           [24]*2 + [18] + [24]*2 + [30]*3 + [24]*9 + [30, 24, 18] + [30]*2 + [18, 30, 24] +
           [30]*2 + [18, 24] + [18]*3 + [24]*3 + [18]*2 + [30] + [24]*2 + [30]*5 + [24]*2 + [30]*6 +
           [24] + [30]*2 + [24] + [30]*6 + [24]*3 + [30]*8 + [18] + [24]*17 + [21] + [30]*8 +
           [21, 18] + [24]*11 + [18, 24] + [18]*5 + [24]*2 + [30]*3 + [24]*2 + [18]*2 + [30]*2 +
           [18] + [24]*2 + [18] + [30]*2 + [18]*3 + [24] + [18]*9 + [30, 18, 24] + [18]*4 + [24])

env = ObstacleTowerEnv(ENV_PATH, worker_id=0)
env.reset()  # Extra reset appears to be important.
env.seed(72)
env.floor(9)
env.reset()
for act in ACTIONS:
    env.step(act)
```

I've also attached a video of what the agent is doing up until the hang:

[hang.zip](https://github.com/Unity-Technologies/obstacle-tower-env/files/3052079/hang.zip)

**Comment by rwightman** [comment]
@unixpickle Not quite the exact same positioning, but the hang I captured was I think the same level and very similar, driving towards that same exit and hanging right before getting there. It'd be nice to get this sorted.

**Comment by VilmantasBal** [comment]
Hey guys,

thanks a lot of this. I will try this script right away. Unfortunately,
although i had running training for more than a week now, i couldn't
reproduce the problem here locally. Perhaps this script will make it.

Kind regards,
Vilmantas

**Comment by VilmantasBal** [comment]
Hey,

just wanted to let you know that i managed to reproduce the issue with a
slight modification of your script. The key part was env.reset() function -
this one was causing the havoc, the level generation has a bug resulting in
agent being destroyed. We are working on a fix.

Kind regards,
Vilmantas

**Comment by miha-skalic** [comment]
@VilmantasBal thanks for preparing the fix. Can you give an estimated time when it will be ready?

Also, are there any workarounds that we could use in the meantime?

**Comment by awjuliani** [comment]
Hi all,

We just pushed a new version of Obstacle Tower (v1.3) which should address these issues: https://github.com/Unity-Technologies/obstacle-tower-env/releases/tag/v1.3. Please let us know if you still see them (or anything else strange) with this new build.

**Comment by unixpickle** [comment]
Still seeing hangs on higher floors (e.g. 11). Might be a different bug, as these hangs seem to happen during a reset(). I'm working on reproducing it.

**Comment by awjuliani** [comment]
Sorry to hear that there are still issues, @unixpickle. If you can pin down a specific seed/floor, that would be very helpful in terms of helping us debug.

**Comment by rwightman** [comment]
@awjuliani @unixpickle with version 1.2 of the environment, I did actually see a hang on a reset the other week. I didn't think much of it then, but that was likely this separate issue. Unfortunately when it happened I wasn't  monitoring for blocking of the reset call, but I am now.

**Comment by unixpickle** [comment]
@awjuliani currently working on reproducing the bug in a deterministic fashion (similar to the script I posted above). However, as reproducing requires running a bunch of episodes with a compute-intensive agent, it will probably be a day or two before I can post a script that reproduces.

**Comment by unixpickle** [comment]
Here's a reproduction of the bug. Like before, this script worked on two different Ubuntu 18.04 machines, and it is deterministic.

It looks like `seed 58, floor 10` is the magic bug-causer here. The other seeds seem to vary between reproductions, but I'll have to verify that. 

```python
from obstacle_tower_env import ObstacleTowerEnv

env = ObstacleTowerEnv('./ObstacleTower/obstacletower', worker_id=0)

env.seed(72)
env.floor(12)
env.reset()
for action in [18, 18, 18, 18, 18, 18, 30, 24, 24, 21, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 24, 24, 6, 6, 6, 6, 6, 6, 6, 6, 30, 30, 30, 30, 30, 18, 24, 24, 24, 6, 6, 6, 6, 6, 6, 24, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 24, 24, 24, 18, 30, 18, 18, 30, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 30, 24, 24, 30, 30, 24, 24, 24, 30, 30, 30, 30, 30, 18, 18, 18, 18, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 18, 30, 18, 18, 18, 18, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 24, 24, 24, 24, 24, 24, 24, 24, 18, 30, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 24, 21, 18, 24, 24, 24, 24, 18, 18, 18, 24, 18, 18, 18, 18, 30, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 18, 18, 30, 30, 30, 30, 30, 30, 12, 12, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 21, 18, 18, 30, 24, 24, 30, 30, 30, 30, 21, 18, 24, 24, 24, 24, 24, 30, 30, 18, 18, 24, 21, 18, 18, 18, 30, 18, 24, 24, 30, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 6, 6, 6, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 18, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 30, 30, 18, 24, 18, 30, 18, 18, 24, 18, 18, 24, 30, 30, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 18, 24, 24, 24, 6, 6, 6, 6, 6, 6, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 18, 30, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 18, 24, 24, 24, 24, 24, 18, 18, 18, 18, 30, 18, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 30, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 24, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 30, 18, 24, 24, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 21, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 18, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 24, 24, 6, 6, 6, 6, 6, 6, 6, 30, 30, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 24, 6, 6, 6, 6, 6, 6, 6, 30, 30, 30, 30, 30, 24, 24, 24, 6, 6, 6, 6, 6, 24, 24, 24, 24, 18, 18, 18, 30, 30, 18, 18, 24, 30, 24, 12, 12, 30, 30, 30, 30, 18, 18, 24, 24, 6, 6, 6, 6, 6, 6, 6, 6, 30, 30, 6, 6, 6, 6, 30, 18, 18, 18, 18, 24, 18, 30, 30, 18, 6, 6, 6, 6, 6, 24, 30, 18, 24, 24, 6, 6, 6, 6, 24, 24, 30, 30, 30, 30, 24, 24, 6, 6, 6, 6, 6, 6, 24, 30, 30, 30, 18, 18, 24, 24, 6, 6, 6, 6, 6, 24, 24, 24, 24, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 18, 18, 18, 24, 24, 18, 18, 18, 18, 18, 30, 30, 30, 18, 18, 24, 24, 18, 6, 6, 18, 18, 18, 18, 18, 18, 18, 30, 30, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 18, 18, 18, 18, 30, 30, 30, 30, 30, 24, 24, 24, 18, 18, 24, 30, 30, 30, 18, 18, 18, 30, 30, 30, 30, 18, 18, 24, 18, 30, 30, 24, 24, 30, 30, 30, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 18, 18, 18, 18, 18, 30, 24, 24, 24, 6, 6, 6, 6, 6, 6, 24, 24, 24, 24, 30, 30, 24, 24, 30, 30, 24, 24, 24, 24, 24, 21, 30, 24, 24, 24, 24, 18, 18, 18, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 21, 21, 18, 18, 18, 24, 18, 18, 18, 18, 30, 18, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 24, 30, 30, 24, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 21, 24, 24, 24, 18, 18, 30, 24, 30, 24, 24, 24, 24, 24, 18, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 18, 30, 24, 24, 18, 18, 30, 30, 30, 30, 30, 30, 18, 18, 21, 18, 24, 30, 18, 18, 24, 24, 6, 6, 6, 24, 24, 24, 18, 24, 24, 24, 24, 21, 18, 30, 18, 18, 24, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 18, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 18, 18, 30, 18, 18, 18, 18, 30, 30, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 24, 21, 24, 24, 24, 24, 18, 18, 30, 30, 24, 24, 24, 18, 24, 24, 30, 30, 30, 24, 24, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 18, 24, 24, 18, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 18, 30, 30, 18, 24, 18, 18, 18, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 30, 30, 24, 24, 30, 30, 18, 18, 30, 30, 30, 18, 24, 24, 24, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 30, 18, 24, 18, 18, 18, 30, 24, 30, 30, 18, 30, 24, 24, 24, 24, 24, 18, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 21, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 30, 30, 24, 24, 21, 18, 18, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 18, 18, 18, 18, 18, 18, 18, 24, 24, 30, 30, 30, 24, 24, 30, 30, 18, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 30, 30, 24, 24, 30, 30, 30, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 24, 24, 18, 18, 30, 30, 18, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 30, 30, 18, 21, 21, 24, 24, 24, 24, 18, 30, 30, 18, 18, 30, 24, 24, 18, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 30, 24, 24, 24, 24, 24, 24, 24, 24, 21, 21, 24, 24, 24, 24, 24, 24, 18, 18, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 30, 30, 30, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 18, 24, 24, 24, 24, 18, 18, 18, 18, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 30, 30, 30, 24, 24, 18, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 30, 30, 18, 18, 24, 24, 24, 18, 18, 18, 18, 30, 30, 24, 24, 24, 24, 18, 18, 30, 30, 30, 30, 30, 30, 30, 18, 18, 18, 30, 30, 30, 30, 30, 30, 30, 30, 18, 24, 18, 24, 30, 30, 18, 18, 18, 24, 24, 18, 30, 30, 30, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 30, 30, 24, 18, 21, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 18, 18, 18, 18, 24, 30, 18, 24, 30, 24, 24, 24, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 18, 18, 21, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 18, 18]:
    env.step(action)

env.seed(58)
env.floor(10)
env.reset()
for action in [18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 24, 30, 30, 30, 30, 30, 30, 30, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 24, 18, 18, 18, 18, 18, 24, 21, 18, 30, 30, 24, 18, 18, 18, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 24, 24, 24, 24, 18, 18, 18, 18, 18, 24, 18, 18, 24, 18, 18, 18, 18, 18, 18, 24, 18, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 6, 6, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 30, 18, 18, 18, 24, 30, 18, 18, 18, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 24, 18, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 30, 30, 18, 18, 24, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 18, 18, 18, 18, 18, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 21, 18, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 18, 30, 18, 18, 24, 24, 18, 18, 18, 18, 18, 30, 18, 6, 6, 6, 6, 6, 24, 24, 24, 24, 24, 24, 24, 21, 30, 18, 18, 18, 18, 30, 18, 18, 24, 24, 18, 18, 30, 30, 30, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 24, 24, 18, 18, 18, 18, 18, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 6, 6, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 24, 24, 30, 30, 30, 30, 18, 24, 24, 6, 6, 6, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 18, 24, 24, 24, 18, 18, 18, 18, 30, 30, 24, 24, 18, 18, 18, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 21, 30, 30, 18, 18, 24, 18, 18, 30, 30, 30, 30, 30, 24, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 18, 18, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 30, 30, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 6, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 18, 18, 18, 18, 24, 18, 18, 18, 30, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 30, 30, 30, 30, 18, 18, 18, 18, 30, 30, 18, 18, 21, 21, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 21, 30, 18, 24, 18, 30, 18, 18, 18, 18, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 18, 30, 30, 24, 24, 24, 24, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 30, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 24, 30, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 24, 24, 24, 6, 6, 6, 6, 6, 6, 6, 6, 30, 30, 30, 30, 30, 30, 18, 24, 24, 6, 6, 6, 6, 6, 6, 6, 18, 18, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 6, 6, 6, 6, 24, 24, 24, 24, 30, 30, 30, 30, 30, 24, 24, 24, 6, 6, 6, 6, 6, 6, 24, 30, 30, 30, 30, 24, 24, 24, 6, 6, 6, 6, 6, 6, 24, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 18, 6, 6, 6, 6, 6, 6, 6, 6, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 6, 6, 24, 24, 24, 24, 24, 24, 30, 30, 30, 30, 18, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 21, 18, 18, 18, 18, 24, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 24, 24, 24, 24, 24, 24, 18, 18, 18, 24, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 24, 24, 24, 6, 6, 6, 6, 6, 6, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 6, 24, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 24, 18, 18, 18, 30, 18, 18, 18, 30, 18, 18, 30, 30, 21, 18, 18, 18, 18, 24, 18, 18, 21, 24, 24, 18, 18, 18, 24, 12, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 18, 30, 18, 21, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 30, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 21, 18, 18, 18, 18, 18, 18, 18, 18, 18, 21, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 30, 18, 30, 30, 30, 18, 24, 24, 6, 6, 6, 6, 6, 6, 6, 6, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 6, 6, 6, 6, 6, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 24, 30, 30, 18, 18, 18, 24, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 18, 24, 24, 30, 24, 24, 24, 24, 24, 24, 18, 18, 18, 18, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 24, 18, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 21, 18, 18, 21, 18, 18, 18, 18, 21, 21, 21, 18, 18, 18, 18, 18, 18, 18, 21, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 24, 18, 18, 18, 18, 30, 18, 18, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 24, 6, 18, 24, 24, 24, 24, 6, 6, 6, 6, 6, 24, 18, 18, 18, 18, 30, 18, 18, 18, 18, 18, 18, 18, 18, 30, 30, 30, 18, 18, 24, 24, 24, 6, 6, 6, 6, 6, 6, 6, 18, 18, 18, 18, 24, 24, 18, 18, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 30, 18, 18, 18, 30, 30, 30, 18, 18, 18, 18, 30, 18, 24, 24, 24, 24, 24, 18, 24, 30, 18, 18, 18, 18, 24, 30, 30, 30, 30, 30, 30, 30, 30, 24, 24, 24, 6, 6, 6, 6, 6, 24, 24, 18, 18, 18, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 18, 18, 18, 18, 30, 18, 24, 24, 18, 18, 30, 30, 18, 24, 24, 24, 30, 30, 30, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 18, 18, 18, 18, 24, 24, 18, 18, 30, 24, 6, 6, 6, 6, 6, 6, 24, 24, 24, 18, 18, 30, 30, 30, 18, 18, 24, 24, 24, 18, 18, 30, 30, 30, 30, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 18, 18, 18, 18, 18, 18, 18, 18, 18, 24, 24, 21, 30, 12, 30, 18, 18, 18, 18, 30, 30, 18, 18, 18, 18, 24, 30, 30, 30, 30, 18, 18, 24, 24, 24, 24, 6, 6, 6]:
    env.step(action)

env.seed(37)
env.floor(11)

print('resetting (this will hang)...')
env.reset()
print('done reset')
env.close()
```

**Comment by awjuliani** [comment]
Hi @unixpickle,

Thanks for reproducing this so quickly. I will take a look. Knowing that it didn't happen in older versions helps narrow down the possible causes by quite a lot.

**Comment by awjuliani** [comment]
Hi @unixpickle 

I finally had time to attempt to reproduce you issue, and am unable to on Mac OS. I ran the script multiple times using the v1.3 binary, and the script was able to finish every time. I will attempt to reproduce on a linux environment next to see whether it may be platform specific.

**Comment by unixpickle** [comment]
@awjuliani I do not have a Mac to test this on. You could try firing up a GCE instance and testing the reproduction there, although I have not tested that.
